### PR TITLE
MIGENG-319: Oracle JDK Workload

### PR DESCRIPTION
**Pull request by m-brophy**
https://issues.redhat.com/browse/MIGENG-319

### CODE REVIEW

**Comment by jonathanvila** [comment]
I'm not sure if I have understood the request on https://issues.redhat.com/browse/MIGENG-319
so, do you need to detect the Java version, or the concrete JVM version and vendor ?
because the rule is checking for the version : 
`                fileNameAndContentsContainSpecifiedStrings(this, "/usr/java/latest/release","JAVA_VERSION=\\\"11")`
but then associating that to Oracle
`addWorkload("Oracle JDK 11")`
but, if the JVM installed is OpenJDK, Zulu, Oracle, Graal, Coretto, etc..... would we detect the same JAVA_VERSION but not belonging to Oracle ?

File: src/main/resources/org/jboss/xavier/analytics/rules/workload/inventory/Workloads.drl [hunk 1]
```diff
@@ -290,3 +290,48 @@ rule "SsaDisabled_System_Services_Not_Present"
         }
 
 end
+
+rule "Workloads_Oracle_JDK_8_On_Linux"
+       when
+           vmWorkloadInventoryModel : VMWorkloadInventoryModel(
+               files != null,
+               files.size() > 0,
+               fileNameAndContentsContainSpecifiedStrings(this, "/usr/java/latest/release","JAVA_VERSION=\\\"1.8")
```
Comment (jonathanvila) [inline src/main/resources/org/jboss/xavier/analytics/rules/workload/inventory/Workloads.drl:299]: The folder would not depend on the different Jdk installed ? I mean that machine could have Java 8 and Java 11 , and then latest would not be 8 but 11

**Comment by jonathanvila** [comment]
Review assumptions : 
* Oracle JDK is installed using this path : /usr/java/latest/release
* only Oracle JDK uses that path ( > 1.6 Jdk )
* the current JDK when having some installed, will always be the "latest"
* the code is only checking for the Oracle JVM
* only considering RHEL as the SO

File: src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java [hunk 2]
```diff
@@ -1236,5 +1236,110 @@ public void testCisco_CallManager() {
...
+        Assert.assertEquals(1, reports.size());
+        WorkloadInventoryReportModel report = reports.get(0);
+        Assert.assertNotNull(report.getWorkloads());
+        Assert.assertEquals(1, report.getWorkloads().size());
+        Assert.assertTrue(report.getWorkloads().stream().anyMatch(workload -> workload.toLowerCase().contains("Oracle JDK 8".toLowerCase())));
+        Assert.assertTrue(report.getSsaEnabled());
+    }
+
+    @Test
+    public void testOracleJDK11OnLinux() {
```
Comment (jonathanvila) [inline src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java:1243]: all 3 tests are 99% identical ..... probably you could extract the common code and refactor them ? to remove duplicated code

**Comment by jonathanvila** [comment]
Tested on CI and apparently doesnt work ( tested with the test payload mentioned in the jira ticket )
Screenshot-20200416170553-362x516

File: src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java [hunk 3]
```diff
@@ -1236,5 +1236,91 @@ public void testCisco_CallManager() {
...
+        facts.put("vmWorkloadInventoryModel", vmWorkloadInventoryModel);
+
+        WorkloadInventoryReportModel workloadInventoryReportModel = new WorkloadInventoryReportModel();
+
+        facts.put("workloadInventoryReportModel",workloadInventoryReportModel);
+
+        Map<String, Object> results = createAndExecuteCommandsAndGetResults(facts);
+
+        Assert.assertEquals(3, results.get(NUMBER_OF_FIRED_RULE_KEY));
+        Utils.verifyRulesFiredNames(this.agendaEventListener, "AgendaFocusForTest", "SsaEnabled_System_Services_Present", ruleNameCheckText);
```
Comment (jonathanvila) [inline src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java]: What do you think moving : 
`Utils.verifyRulesFiredNames(this.agendaEventListener, "AgendaFocusForTest", "SsaEnabled_System_Services_Present", ruleNameCheckText);`
to
```
String[] rulesNamesX = {"AgendaFocusForTest", "SsaEnabled_System_Services_Present", ruleNameCheckText};
Utils.verifyRulesFiredNames(this.agendaEventListener,Arrays.stream(rulesNamesX).filter(e-> e != null).toArray(String[]::new));
```
In order to allow receiving the parameter "null"

Also doing the Asser.assertEquals(3, results.....) to be Assert.assertEquals(rulesNamesX.size)

File: src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java [hunk 4]
```diff
@@ -1236,5 +1236,91 @@ public void testCisco_CallManager() {
...
+        Assert.assertEquals(1, reports.size());
+        WorkloadInventoryReportModel report = reports.get(0);
+        Assert.assertNotNull(report.getWorkloads());
+        Assert.assertEquals(1, report.getWorkloads().size());
+        Assert.assertTrue(report.getWorkloads().stream().anyMatch(workload -> workload.toLowerCase().contains(workloadCheckText.toLowerCase())));
+        Assert.assertTrue(report.getSsaEnabled());
+    }
+
+    @Test
+    public void testExcessEscapeCharsNoOracleJDKDetected()
```
Comment (jonathanvila) [inline src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java:1297]: With my previous comment you could refactor this method and do a call to :
`testOracleJDK("JAVA_VERSION=\"13", null, "Oracle JDK 13");`
Extracting the last lines of the method ( asserts ) to the test body, receiving the `reports` collection as the return of the method.

Comment (jonathanvila) [reply src/test/java/org/jboss/xavier/analytics/rules/workload/inventory/WorkloadsTest.java:1293]: test

**Comment by m-brophy** [comment]
I've done most of what you suggested - had to alter the number of rules fired test slightly, and I didn't return report from the common method and do asserts in each test - I thought it would be better to keep the whole common test method encapsulated

**Review by jonathanvila — APPROVED** [review]

**Pull request merged**